specs/marketplace.md

```diff
@@ -0,0 +1,374 @@
+the freed slot MAY also participate in the data repair, but my refilling the slot it **won't** recover its original collateral
+and needs to submit new collateral with the `fillSlot()` call.
+
+The repair process is the same as with the filling slots. If the original slot's dataset is not present in the network
+the node MAY use the erasure coding to reconstruct the original slot's dataset. 
+As this requires retrieving more data of the dataset from the network, the node that will successfully repair
+slot by filling the freed slot will be granted an additional reward. (**TODO: Implementation**)
+
+The repair process is then as follows:
+
+1. Node detects `SlotFreed` event and decide to repair it.
+1. Node MUST download the required chunks and MUST use the [Reed-Soloman algorithm](https://hackmd.io/FB58eZQoTNm-dnhu0Y1XnA) to reconstruct the original slot's data.
+1. Node MUST generate proof over the reconstructed data.
+1. Node MUST submit transaction with call to `fillSlot()` with the same parameters and collateral allowance as described in the [Filling slot](#filling-slot).
+
+### Collecting funds
+
+A Storage Provider node SHOULD monitor Requests and slots it hosts. 
+
+When a node slot's Requests reaches states `Cancelled`, `Finished` or `Failed` it SHOULD call the contract's `freeSlot(slotId)` function.
+These states can be detected using:
+
+ - `Cancelled` state MAY be detected by setting timeout using `expiry` **and** not detecting `RequestFulfilled(requestId)` event. There is also `RequestCancelled` event emitted, but the node SHOULD NOT use it for asserting expiry as it is not guaranteed to be emitted at the time of expiry.
+ - `Finished` state MAY be detected by setting timeout specified from function `getRequestEnd(requestId)`.
+ - `Failed` state MAY be detected by listening to the `RequestFailed(requestId)` event emitted.
+
+For each of these states, different funds are collected:
+
+- For `Cancelled` the collateral is returned together with proportional payout based on time that the node actually hosted the dataset before expiry was reached.
+- For `Finished` the full reward for hosting the slot together with collateral is gathered.
+- For `Failed` no funds are collected as reward is returned to the client and collateral is burned, but this call removes the slot from the `mySlots()` tracking.
+
+## Validator role
+
+Validator role represents nodes that help to verify that the Storage Provider nodes submit proofs when they are required.  
+The decision maker around if proof was missed or not is the smart contract, validator only triggers the decision-making
+function on the smart contract. This is because in a blockchain we cannot act on things that **do not happen** and somebody needs to create a transaction 
+in order for the smart contract to act on it. The validator nodes get then rewarded for each time they correctly
+mark proof as missing.
```

Comment:
```suggestion
The Validator role represents nodes that help to verify that SP nodes submitted proofs when they were required.  
The smart contract decides whether or not a proof was missed, while the validator triggers the decision-making
function in the smart contract. This is because in a blockchain, a contract cannot change its state without a transaction and gas initiating the state change. The validator nodes then get rewarded for each time they correctly
mark a proof as missing.
```